src/codegen/operator/table_scan_translator.cpp

```diff
@@ -6,156 +6,304 @@
+  };
+
+  // Execute serially
+  GetPipeline().RunSerial(producer);
 }
 
-// Get the stringified name of this scan
-std::string TableScanTranslator::GetName() const {
-  std::string name = "Scan('" + GetTable().GetName() + "'";
-  auto *predicate = GetScanPlan().GetPredicate();
-  if (predicate != nullptr && predicate->IsSIMDable()) {
-    name.append(", ").append(std::to_string(Vector::kDefaultVectorSize));
-  }
-  name.append(")");
-  return name;
+void TableScanTranslator::ProduceParallel() const {
+  CodeGen &codegen = GetCodeGen();
+
+  // The input arguments
+  const storage::DataTable &table = *GetScanPlan().GetTable();
+
+  // The use RuntimeFunctions::ExecuteTableScan() to launch a parallel scan.
+  // We pass in the database and table IDs to scan the correct table.
+  auto *dispatcher =
+      RuntimeFunctionsProxy::ExecuteTableScan.GetFunction(codegen);
+  std::vector<llvm::Value *> dispatch_args = {
+      codegen.Const32(table.GetDatabaseOid()), codegen.Const32(table.GetOid())};
+
+  // Our function needs to know the start and stop tile groups to scan
+  std::vector<llvm::Type *> pipeline_arg_types = {codegen.Int64Type(),
+                                                  codegen.Int64Type()};
+
+  // Parallel production
+  auto producer = [this, &codegen](ConsumerContext &ctx,
+                                   const std::vector<llvm::Value *> params) {
+    PELOTON_ASSERT(params.size() == 2);
+
+    // Load the table pointer
+    auto *table_ptr = LoadTablePtr(codegen);
+
+    // The selection vector for the scan
+    auto *i32_type = codegen.Int32Type();
+    auto vec_size = Vector::kDefaultVectorSize.load();
+    auto *raw_vec = codegen.AllocateBuffer(i32_type, vec_size, "scanPosList");
+    Vector position_list{raw_vec, vec_size, i32_type};
+
+    // zonemap
+    auto predicate = const_cast<expression::AbstractExpression *>(
+        GetScanPlan().GetPredicate());
+    llvm::Value *predicate_ptr = codegen->CreateIntToPtr(
+        codegen.Const64((int64_t)predicate),
+        AbstractExpressionProxy::GetType(codegen)->getPointerTo());
+    size_t num_preds = 0;
+
+    auto *zone_map_manager = storage::ZoneMapManager::GetInstance();
+    if (predicate != nullptr && zone_map_manager->ZoneMapTableExists()) {
+      if (predicate->IsZoneMappable()) {
+        num_preds = predicate->GetNumberofParsedPredicates();
+      }
+    }
+
+    // Scan the given range of the table
+    ScanConsumer scan_consumer{ctx, GetScanPlan(), position_list};
+    table_.GenerateScan(codegen, table_ptr, params[0], params[1], vec_size,
+                        predicate_ptr, num_preds, scan_consumer);
+  };
+
+  // Execute parallel
+  auto &pipeline = GetPipeline();
+  pipeline.RunParallel(dispatcher, dispatch_args, pipeline_arg_types, producer);
 }
 
-// Table accessor
-const storage::DataTable &TableScanTranslator::GetTable() const {
-  return *scan_.GetTable();
+void TableScanTranslator::Produce() const {
+  if (GetPipeline().IsParallel()) {
```

Comment:
Correct. Each operator can choose whether they want to be parallel or not. Unless an operator in the pipeline specifically calls `Pipeline::SetSerial`, the decision for the parallelism of the pipeline is made in `Pipeline::MarkSource`. The decision takes into account:
1. Whether parallel execution is enabled.
2. Whether the consumer supports parallel execution.
3. The preference of parallelism of all operators in the pipeline.

I think the comments are clear, but let me know otherwise!